R/cnd-last.R

```diff
@@ -0,0 +1,117 @@
+#' Display last warnings
+#'
+#' @description
+#'
+#' * `last_warnings()` and `last_messages()` return a list of all
+#'   warnings and messages that occurred during the last top-level R
+#'   command.
+#'
+#' * `last_warning()` and `last_message()` return the very last ones.
+#'
+#' By default the warnings and messages are printed with a simplified
+#' backtrace, like [last_error()]. Use `summary()` to print the
+#' conditions with a full backtrace.
+#'
+#' @param n How many warnings or messages to display. Defaults to all.
+#'
+#' @examples
+#' if (FALSE) {
+#'   global_entrace()
+#'   f <- function() { warning("foo"); g() }
+#'   g <- function() { warning("bar", immediate. = TRUE); h() }
+#'   h <- function() warning("baz")
+#'
+#'   f()
+#'
+#'   # Simplified backtraces
+#'   last_warnings()
+#'
+#'   # Full backtraces
+#'   summary(last_warnings())
+#' }
+#' @export
+last_warnings <- function(n = NULL) {
+  out <- new_list_of_conditions(the$last_warnings)
+  tail(out, n = n %||% length(out))
+}
+#' @rdname last_warnings
+#' @export
+last_messages <- function(n = NULL) {
+  out <- new_list_of_conditions(the$last_messages)
+  tail(out, n = n %||% length(out))
+}
+
+on_load({
+  the$last_top_frame <- NULL
+  the$last_warnings <- list()
+  the$last_messages <- list()
+})
+
+push_warning <- function(cnd) {
+  top <- obj_address(sys.frame(1))
```

Comment:
I've added this comment:

```r
# We collect warnings/messages in a list as long as the first frame on
# the call stack has the same address. If there is a new address, it
# must be that a new top-level R command is running and we start a new
# list. This heuristic is technically not 100% correct. We might be
# very unlucky: If (a) a GC occur between two commands (b) the first
# frame on the stack reuses the same address as the last first frame
# on the stack, then we'll wrongly keep collecting warnings instead of
# starting anew.
```